test/Streamly/Test/Prelude/Serial.hs

```diff
@@ -106,55 +107,207 @@ splitOnSuffixSeq = do
+
+    nonSepElem :: Gen a
+    nonSepElem = suchThat arbitrary (/= sep)
+
+    listWithSep :: Gen [a]
+    listWithSep = listOf $ frequency [(3, arbitrary), (1, elements [sep])]
+
+    listWithoutSep :: Gen [a]
+    listWithoutSep = vectorOf 4 nonSepElem
+
+    listsWithoutSep :: Gen [[a]]
+    listsWithoutSep = listOf listWithoutSep
+
+    listsWithoutSep1 :: Gen [[a]]
+    listsWithoutSep1 = listOf1 listWithoutSep
+
+    intercalateSplitEqId i =
+        let name =
+                "intercalate . splitOnSeq == id ("
+                    <> show i <> " element separator)"
+         in prop name
+                $ forAll listWithSep
+                $ \xs -> withMaxSuccess maxTestCount $ monadicIO $ testCase xs
+
+        where
+
+        testCase xs = do
+            ys <- splitOnSeq_ xs (replicate i sep)
+            zs <-
+                IS.toList
+                    $ IS.intercalate (replicate i sep) UF.fromList
+                    $ IS.fromList ys
+            let zs1 = intercalate (replicate i sep) ys
+            listEquals (==) zs xs
+            listEquals (==) zs1 xs
+
+    intercalateSplitEqIdNoSepEnd i =
+        let name =
+                "intercalate . splitOnSuffixSeq_ . (++ [x \\= sep]) == id ("
+                    <> show i <> " element separator)"
+         in prop name
+                $ forAll ((,) <$> listWithSep <*> nonSepElem)
+                $ \(xs_, nonSep) -> do
+                      let xs = xs_ ++ [nonSep]
+                      withMaxSuccess maxTestCount $ monadicIO $ testCase xs
+
+        where
+
+        testCase xs = do
+            ys <- splitOnSuffixSeq_ xs (replicate i sep)
+            zs <-
+                IS.toList
+                    $ IS.intercalate (replicate i sep) UF.fromList
+                    $ IS.fromList ys
+            let zs1 = intercalate (replicate i sep) ys
+            listEquals (==) zs xs
+            listEquals (==) zs1 xs
+
+    concatSplitIntercaEqConcat splitter intercalater i =
+        let name =
+                "concat . splitter . intercalater == "
+                    <> "concat ("
+                    <> show i <> " element separator/possibly empty list)"
+         in prop name
+                $ forAll listsWithoutSep
+                $ \xss -> do
+                      withMaxSuccess maxTestCount
+                          $ monadicIO
+                          $ do
+                              let xs = intercalater (replicate i sep) xss
+                              ys <- splitter xs (replicate i sep)
+                              listEquals (==) (concat ys) (concat xss)
+
+    concatSplitIntercaEqConcatS splitter intercalater i =
+        let name =
+                "concat . splitter . S.intercalater == "
+                    <> "concat ("
+                    <> show i <> " element separator/possibly empty list)"
+         in prop name
+                $ forAll listsWithoutSep
+                $ \xss -> withMaxSuccess maxTestCount $ monadicIO $ testCase xss
```

Comment:
Why can't we combine `concatSplitIntercaEqConcatS` with `concatSplitIntercaEqConcat`?